addons/bestja_detailed_reports/models.py

```diff
@@ -0,0 +1,215 @@
+    ]
+
+
+class DetailedReport(models.Model):
+    _name = 'bestja.detailed_report'
+    _inherit = [
+        'protected_fields.mixin',
+        'ir.needaction_mixin',
+        'message_template.mixin',
+    ]
+    _order = 'write_uid desc'
+    STATES = [
+        ('sent', "wysłany"),
+        ('accepted', "zaakceptowany"),
+        ('draft', "szkic"),
+        ('rejected', "odrzucony"),
+    ]
+
+    project = fields.Many2one(
+        'bestja.project',
+        required=True,
+        string="Projekt",
+        domain=lambda self: [
+            ('parent.use_detailed_reports', '=', True),
+            ('detailed_reports', '=', False),
+            '|',
+            ('manager', '=', self.env.user.id),
+            ('organization.coordinator', '=', self.env.user.id)],
+        ondelete='restrict',
+    )
+    organization = fields.Many2one(
+        'organization',
+        string="Organizacja",
+        related='project.organization',
+    )
+    name = fields.Char(string="Nazwa projektu", related="project.name")
+    dates = fields.Char(string="Termin", compute="compute_project_dates", store=True)
+    state = fields.Selection(STATES, default='draft', string="Status:")
+    report_entries = fields.One2many('bestja.report_entry', inverse_name='detailed_report', string="Produkt")
+    tonnage = fields.Float(string="Tonaż (kg)", compute="compute_report_tonnage", store=True)
+    parent_project = fields.Many2one(
+        'bestja.project',
+        string="Projekt nadrzędny",
+        related='project.parent',
+        store=True,
+    )
+    interested_organization = fields.Many2one(
+        'organization',
+        string="Zainteresowane organizacje",
+        compute="compute_interested_organization",
+        store=True,
+    )
+    user_can_moderate = fields.Boolean(compute="compute_user_can_moderate")
+
+    @api.one
+    @api.depends('parent_project', 'project')
+    def compute_interested_organization(self):
+        level = self.project.organization.level
+        if level <= 1:
+            self.interested_organization = self.project.organization.id
+        else:
+            self.interested_organization = self.parent_project.organization.id
+
+    @api.one
+    @api.depends('report_entries')
+    def compute_report_tonnage(self):
+        """
+        For showing the sum of kg of products
+        """
+        self.tonnage = 0.0
+        for q in self.report_entries:
+            self.tonnage += q.tonnage
+
+    @api.one
+    @api.depends('parent_project')
+    def compute_user_can_moderate(self):
+        """
+        Is current user authorized to moderate (accept/reject) the detailed_report?
+        """
+        self.user_can_moderate = (
+            self.parent_project.manager.id == self.env.user.id or
+            self.parent_project.organization.coordinator.id == self.env.user.id
+        )
+
+    @api.one
+    @api.constrains('report_entries')
+    def check_report_entries(self):
```

Comment:
Don't you already have an SQL constraint that already covers this case (in a much more robust way)? On line 42.